Prefer setting struct fields by passing them to `new`

fingolfin opened the pull request:
Some people prefer writing constructors as e.g.

    mutable struct Foo
      coeffs::Vector{Int}
      exps::Vector{Int}
      optional::Vector{Int}
      function Foo(x::Vector{Int}, y::Vector{Int})
        f = new()
        f.coeffs = x
        f.exps = y
        return f
      end
    end

instead of

    mutable struct Foo
      coeffs::Vector{Int}
      exps::Vector{Int}
      optional::Vector{Int}
      function Foo(x::Vector{Int}, y::Vector{Int})
        return new(x, y)
      end
    end

The primary arguments are readability and safety: this way it is immediately
clear which expression is assigned to which field (you don't need to count the
arguments to `new` to determine this). And this then reduces the risk of
mistakes where one accidentally assigns a constructor argument to the wrong
fields (if the types match, otherwise of course a runtime error will be thrown
anyway).

However, this also has performance ramifications: consider a simple helper function

    bar(foo::Foo) = length(foo.coeffs)

In the second example, Julia is able to deduce that `foo.coeffs` will always
be initialized, and generates optimal code machine code (edited slightly for
readability):

    julia> foo = Foo([1,2],[3,4]); @code_native bar(foo)
        .section	__TEXT,__text,regular,pure_instructions
        movq	(%rdi), %rax
        movq	8(%rax), %rax
        retq

But in the former case, Julia is not able to do that, and thus generates code
which is far less efficient:

    julia> foo = Foo([1,2],[3,4]); @code_native bar(foo)
        .section	__TEXT,__text,regular,pure_instructions
        pushq	%rax
        movq	(%rdi), %rax
        testq	%rax, %rax
        je	L15
        popq	%rcx
        retq
    L15:
        movabsq	$jl_throw, %rax
        movabsq	$jl_system_image_data, %rdi
        callq	*%rax
        nopw	%cs:(%rax,%rax)

The difference does not matter if access is rare, but it can accumulate
quickly if one e.g. pays it for access to entries of a matrix and then
performs it millions of times.

This patch thus changes some uses of `new` to try and set fields by
passing them as arguments directly. To get at least somewhat better
readability, we ensure that the names of the arguments passed to `new`
match the field names. So to continue the initial example, it is better
to write

    mutable struct Foo
      coeffs::Vector{Int}
      exps::Vector{Int}
      optional::Vector{Int}
      function Foo(coeffs::Vector{Int}, exps::Vector{Int})
        return new(coeffs, exps)
      end
    end

For even more safety, one can also use the Base.@kwdef macro and then
reimplement existing inner constructors using the keyword-args based
constructor provided by the macro. Note that this does not allow for undefined
fields, but one can replace fields `x::T` which currently can be `undef` by
`x::Union{Nothing,T} = nothing` for a similar but safer effect:

    Base.@kwdef mutable struct Foo
      coeffs::Vector{Int}
      exps::Vector{Int}
      optional::Union{Nothing,Vector{Int}} = nothing
    end
    function Foo(coeffs::Vector{Int}, exps::Vector{Int})
      return Foo(; coeffs, exps)
    end

thofma commented:
I don't think the check whether something is defined is of relevance performance wise for these types. So I don't see the big need to refactor this code?

I am also not too happy introducing `Union{Nothing,}` just to replace `isdefined(...)` by `=== nothing`. I would guess that this makes Oscar compilation times go up even more.

tthsqe12 commented:
The last time I looked at this with julia 1.6 indeed undefined references were just nullptrs and julia only skips the nullptr check on a member when it can prove no constructor can leave it undefined. (You can crash julia by giving it an unexcepted undefined reference). Unnecessary nullptr checks are silly but will be predicted away. I would say from now on let's use `new(a,b)` instead of `new(); .a = ; .b = ` but if refactoring old code is too much work, it is not a big deal to leave these in there.

tthsqe12 commented:
@thofma What are you talking about with this Union{Nothing ? I don't think anyone is suggesting replacing the elegant possibly-undefined member solution to the bivalue problem? Anyhow, these nullptr checks are really everywhere due to the design of the language (note also the 1-based indexing generating "inefficient" code). Since matrix access requires a minimum of two bounds checks and one nullptr check, at some point, we have to throw our hands up and say "who cares" :).

```julia
julia> mutable struct foo; a::Int128; end;

julia> code_native(getindex, (Vector{foo}, Int), syntax=:intel)
	.text
; ┌ @ array.jl:805 within `getindex'
	push	rbp
	mov	rbp, rsp
	lea	rax, [rsi - 1]
	cmp	rax, qword ptr [rdi + 8]
	jae	L32                     ; bound check
	mov	rax, qword ptr [rdi]
	mov	rax, qword ptr [rax + 8*rsi - 8]
	test	rax, rax
	je	L66                     ; undef check
	mov	rsp, rbp
	pop	rbp
	ret

julia> code_native(getindex, (Matrix{foo}, Int, Int), syntax=:intel)
	.text
; ┌ @ array.jl:806 within `getindex'
	push	rbp
	mov	rbp, rsp
	mov	rax, qword ptr [rdi + 24]
	lea	r8, [rsi - 1]
	cmp	r8, rax
	jae	L51                     ; bound check
	lea	rcx, [rdx - 1]
	cmp	rcx, qword ptr [rdi + 32]
	jae	L51                     ; bound check
	imul	rax, rcx
	mov	rcx, qword ptr [rdi]
	add	rax, r8
	mov	rax, qword ptr [rcx + 8*rax]
	test	rax, rax                ; undef check
	je	L89
	mov	rsp, rbp
	pop	rbp
	ret
```

tthsqe12 commented on src/Rings/mpoly.jl:
```diff
@@ -483,35 +476,32 @@ export lex, deglex, degrevlex, weights, MonomialOrdering, singular
 
-  function BiPolyArray(a::Vector{T}; keep_ordering::Bool = true, isGB::Bool = false) where {T <: MPolyElem}
+  function BiPolyArray(O::Vector{T}; keep_ordering::Bool = true, isGB::Bool = false) where {T <: MPolyElem}
     return BiPolyArray(parent(a[1]), a; keep_ordering = keep_ordering,
```
```suggestion
    return BiPolyArray(parent(O[1]), O; keep_ordering = keep_ordering,
```

tthsqe12 commented:
> For even more safety, one can also use the Base.@kwdef macro and then
reimplement existing inner constructors using the keyword-args based
constructor provided by the macro. Note that this does not allow for undefined
fields, but one can replace fields x::T which currently can be undef by
x::Union{Nothing,T} = nothing for a similar but safer effect:

Can see this being useful if we would like the ability to *reset* a member to nullptr. For example, if ideals are mutable(?), then if the oscar data changes, the singular data needs to be nullified, and vice versa. As far as I know, with the `undef` approach, it is not possible to set a defined member back to `undef`.

fingolfin committed:
Prefer setting struct fields by passing them to `new`

Some people prefer writing constructors as e.g.

    mutable struct Foo
      coeffs::Vector{Int}
      exps::Vector{Int}
      optional::Vector{Int}
      function Foo(x::Vector{Int}, y::Vector{Int})
        f = new()
        f.coeffs = x
        f.exps = y
        return f
      end
    end

instead of

    mutable struct Foo
      coeffs::Vector{Int}
      exps::Vector{Int}
      optional::Vector{Int}
      function Foo(x::Vector{Int}, y::Vector{Int})
        return new(x, y)
      end
    end

The primary arguments are readability and safety: this way it is immediately
clear which expression is assigned to which field (you don't need to count the
arguments to `new` to determine this). And this then reduces the risk of
mistakes where one accidentally assigns a constructor argument to the wrong
fields (if the types match, otherwise of course a runtime error will be thrown
anyway).

However, this also has performance ramifications: consider a simple helper function

    bar(foo::Foo) = length(foo.coeffs)

In the second example, Julia is able to deduce that `foo.coeffs` will always
be initialized, and generates optimal code machine code (edited slightly for
readability):

    julia> foo = Foo([1,2],[3,4]); @code_native bar(foo)
        .section	__TEXT,__text,regular,pure_instructions
        movq	(%rdi), %rax
        movq	8(%rax), %rax
        retq

But in the former case, Julia is not able to do that, and thus generates code
which is far less efficient:

    julia> foo = Foo([1,2],[3,4]); @code_native bar(foo)
        .section	__TEXT,__text,regular,pure_instructions
        pushq	%rax
        movq	(%rdi), %rax
        testq	%rax, %rax
        je	L15
        popq	%rcx
        retq
    L15:
        movabsq	$jl_throw, %rax
        movabsq	$jl_system_image_data, %rdi
        callq	*%rax
        nopw	%cs:(%rax,%rax)

The difference does not matter if access is rare, but it can accumulate
quickly if one e.g. pays it for access to entries of a matrix and then
performs it millions of times.

This patch thus changes some uses of `new` to try and set fields by
passing them as arguments directly. To get at least somewhat better
readability, we ensure that the names of the arguments passed to `new`
match the field names. So to continue the initial example, it is better
to write

    mutable struct Foo
      coeffs::Vector{Int}
      exps::Vector{Int}
      optional::Vector{Int}
      function Foo(coeffs::Vector{Int}, exps::Vector{Int})
        return new(coeffs, exps)
      end
    end

For even more safety, one can also use the Base.@kwdef macro and then
reimplement existing inner constructors using the keyword-args based
constructor provided by the macro. Note that this does not allow for undefined
fields, but one can replace fields `x::T` which currently can be `undef` by
`x::Union{Nothing,T} = nothing` for a similar but safer effect:

    Base.@kwdef mutable struct Foo
      coeffs::Vector{Int}
      exps::Vector{Int}
      optional::Union{Nothing,Vector{Int}} = nothing
    end
    function Foo(coeffs::Vector{Int}, exps::Vector{Int})
      return Foo(; coeffs, exps)
    end

fingolfin commented:
I've dropped the `BiPolyArray` changes for now, will put them into another PR. Assuming tests pass, I hope it'll be OK to merge this now trivial PR :-)
